keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md

```diff
@@ -0,0 +1,296 @@
+          * [Current kubelet](#current-kubelet)
+          * [Current kube-proxy](#current-kube-proxy)
+          * [Current kube-scheduler](#current-kube-scheduler)
+          * [Current kube-controller-manager](#current-kube-controller-manager)
+          * [Current kube-apiserver](#current-kube-apiserver)
+          * [Current cloud-controller-manager](#current-cloud-controller-manager)
+       * [Goals](#goals)
+       * [Non-goals](#non-goals)
+       * [Related proposals and further references](#related-proposals-and-further-references)
+    * [Proposal](#proposal)
+       * [Migration strategy per component or k8s.io repo](#migration-strategy-per-component-or-k8sio-repo)
+          * [k8s.io/apimachinery changes](#k8sioapimachinery-changes)
+          * [k8s.io/apiserver changes](#k8sioapiserver-changes)
+          * [kubelet changes](#kubelet-changes)
+          * [kube-proxy changes](#kube-proxy-changes)
+          * [kube-scheduler changes](#kube-scheduler-changes)
+          * [k8s.io/controller-manager changes](#k8siocontroller-manager-changes)
+          * [kube-controller-manager changes](#kube-controller-manager-changes)
+          * [cloud-controller-manager changes](#cloud-controller-manager-changes)
+          * [kube-apiserver changes](#kube-apiserver-changes)
+       * [Timeframe and Implementation Order](#timeframe-and-implementation-order)
+       * [OWNERS files for new packages and repos](#owners-files-for-new-packages-and-repos)
+
+## Summary
+
+Currently all ComponentConfiguration API types are in the core Kubernetes repo, which make them practically inaccessible to for any third-party tool,
+and blocks e.g. kubeadm to move out of the core repo. This KEP is about creating new staging repos, `k8s.io/{component}`, which will host the external
+types of the core components’ ComponentConfig in a top-level `config/` package. Internal types will *eventually* be stored in
+`k8s.io/{component}/pkg/apis/config` (but a non-goal for this KEP). Shared types will go to `k8s.io/{apimachinery,apiserver,controller-manager}/pkg/apis/config`.
+
+### The current state of the world
+
+#### Current kubelet
+
+* **Package**: [k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig](https://github.com/kubernetes/kubernetes/blob/release-1.11/pkg/kubelet/apis/kubeletconfig/types.go)
+* **GroupVersionKind:** `kubelet.config.k8s.io/v1beta.KubeletConfiguration`
+* **Supports** reading **config from file** with **flag precedence**, **well-tested**.
+
+#### Current kube-proxy
+
+* **Package**: [k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig](https://github.com/kubernetes/kubernetes/blob/release-1.11/pkg/proxy/apis/kubeproxyconfig/types.go)
+* **GroupVersionKind**: `kubeproxy.config.k8s.io/v1alpha1.KubeProxyConfiguration`
+* **Supports** reading **config from file**, **without flag precedence**, **not tested**.
+* This API group embeds `ClientConnectionConfiguration`, which is **wrong**
```

Comment:
For things that are "wrong," it would be helpful to cross-reference a related issue.

Reply:
It's less about "wrong" here. Those types are duplicated in each API group instead of being shared. It's something we should unify. I updated the text.